Add NoQEngine to QEngine and refactor the name of set/get qengine

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26331 [quant] Add QuantFusion to graph_executor
* **#26330 Add NoQEngine to QEngine and refactor the name of set/get qengine**

Summary:
att

Test Plan:
.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17464904](https://our.internmc.facebook.com/intern/diff/D17464904)

jerryzh168 committed:
Add NoQEngine to QEngine and refactor the name of set/get qengine

Summary:
att

Test Plan:
.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add NoQEngine to QEngine and refactor the name of set/get qengine"


Summary:
att

Test Plan:
.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

dzhulgakov commented on aten/src/ATen/Context.cpp:
```diff
@@ -95,12 +95,28 @@ bool Context::hasLAPACK() const {
 
-void Context::setPreferredQuantizedEngine(at::QEngine e) {
-  quantized_engine = e;
+void Context::setQEngine(at::QEngine e) {
```
I'd also add Context::supported_qengines which returns a list of supported ones

jerryzh168 replied:
is this for users to know what are the supported qengines so they can choose one?

dzhulgakov commented on torch/backends/quantized/__init__.py:
```diff
@@ -22,7 +22,7 @@ def __init__(self, m, name):
         return self.m.__getattribute__(attr)
 
-    engine = ContextProp(torch._C._get_preferred_engine, torch._C._set_preferred_engine)
+    engine = ContextProp(torch._C._get_qengine, torch._C._set_qengine)
```
btw, I was chatting with Greg that polluting the main namespace with fbgemm/qnnpack is not nice. We could just expose it to python via strings (or None for empty). But it can be done in a separate diff

dzhulgakov reviewed (commented):
which logic do you want to enable? if qengine is set to None - is it allowing to choose either one or does it mean "disable quantization"? if the latter - why is it useful? if the former - can it be just optional<QEngine>?

supriyar commented on aten/src/ATen/Context.cpp:
```diff
@@ -95,12 +95,28 @@ bool Context::hasLAPACK() const {
+      quantized_engine = e;
+      break;
+    case at::kFBGEMM:
+#ifdef FBGEMM
```
Shouldn't this be USE_FBGEMM?

supriyar commented on aten/src/ATen/Context.cpp:
```diff
@@ -95,12 +95,28 @@ bool Context::hasLAPACK() const {
+      break;
+#endif
+    case at::kQNNPACK:
+#ifdef QNNPACK
```
The macro we use for pytorch is USE_PYTORCH_QNNPACK

supriyar commented on aten/src/ATen/Context.cpp:
```diff
@@ -95,12 +95,28 @@ bool Context::hasLAPACK() const {
+      break;
+#endif
+    default:
+      TORCH_CHECK(false, "quantized engine ", toString(e), "is not supported");
```
Should we set an engine by default based on the enabled macros? If the user doesn't call `setQEngine` then does that mean we won't run on any quantized backend?

jerryzh168 commented:
@dzhulgakov see next PR for usage, I want to enable QuantFusion pass only when some quantized engine is set by user. so NoQEngine means quantized kernel is disabled.

@supriyar I think we can setEngine in some initialization code, so that it will check whether that engine is compiled or not as well.

jerryzh168 committed:
Update on "Add NoQEngine to QEngine and refactor the name of set/get qengine"


Summary:
att

Test Plan:
.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

dzhulgakov commented on aten/src/ATen/Context.h:
```diff
@@ -125,7 +126,13 @@ class CAFFE2_API Context {
   bool enabled_mkldnn = true;
-  at::QEngine quantized_engine = at::QEngine::FBGEMM;
+  at::QEngine quantized_engine =
+#ifdef USE_PYTORCH_QNNPACK
```
both fbgemm and qnnpack can be present on server side - we should give fbgemm the priority

dzhulgakov reviewed (commented):
I saw the next diff, but I don't understand under which circumstances you do not want to run QuantFusion pass at all assuming that at lease one engine is compiled in. It feels like you're looking for just a flag "at least one engine is present" not really "preferred one"

jerryzh168 commented on aten/src/ATen/Context.cpp:
```diff
@@ -95,12 +95,28 @@ bool Context::hasLAPACK() const {
+      quantized_engine = e;
+      break;
+    case at::kFBGEMM:
+#ifdef USE_FBGEMM
```
we do throw, see default branch

jerryzh168 replied:
oh, missing break

jerryzh168 committed:
Update on "Add NoQEngine to QEngine and refactor the name of set/get qengine"


Summary:
att

Test Plan:
.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
> I saw the next diff, but I don't understand under which circumstances you do not want to run QuantFusion pass at all assuming that at lease one engine is compiled in. It feels like you're looking for just a flag "at least one engine is present" not really "preferred one"

what I need from next diff is whether user has enabled quantized engine, if quantized engine is NoQEngine then we don't want to fuse because the output graph will not be runnable.

maybe I should check `qEngine() != kNoQEngine` instead.

supriyar commented on aten/src/ATen/Context.cpp:
```diff
@@ -95,12 +95,32 @@ bool Context::hasLAPACK() const {
 }
 
-at::QEngine Context::preferredQuantizedEngine() const {
+std::vector<at::QEngine> Context
```
Forgot to delete?

jerryzh168 replied:
right..

supriyar reviewed (approved):
LGTM!

jerryzh168 committed:
Update on "Add NoQEngine to QEngine and refactor the name of set/get qengine"


Summary:
att

Test Plan:
.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ljk53 commented:
This breaks android CI on master :)

supriyar commented:
I think you need to update qconv.cpp and qlinear.cpp as well.

jerryzh168 commented:
why is this not caught by CI?

jerryzh168 commented:
Some checks were not successful
`1 errored, 4 failing, 1 pending, and 59 successful checks`
the 4 failing ones looks like noise.
